File: skills/java-superpowers-contract/scripts/skill_bridge.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Skill Bridge (Python 版)
连接 java-mysql-query 与 java-superpowers-contract 的桥接工具：
将 DatabaseQuery 输出自动转换为审计报告生成器输入。
用法：
  python skill_bridge.py --analyze-result analyze_result.json --audit-output audit_input.json
  python skill_bridge.py --db mydb --tables user order --audit-format markdown --output combined_report.md
"""
import json, sys, os, argparse, datetime
from pathlib import Path
from database_query import MySQLQuery

QUALITY_WARN_THRESHOLDS = {"nullRatio": 0.2, "emptyStringRatio": 0.3, "sentinelValueRatio": 0.1}

def convert_analyze_to_audit(analyze_data, session_id=None, skills=None, tools=None):
    if isinstance(analyze_data, str):
        with open(analyze_data) as f: analyze_data = json.loads(f.read())
    analysis = analyze_data.get("analysis", analyze_data)
    columns = analysis.get("columns", [])
    table_name = analysis.get("table", "unknown")
    quality_issues = []
    for col in columns:
        warnings = []
        col_name = col.get("name", "unknown")
        nr = col.get("nullRatio", 0)
        esr = col.get("emptyStringRatio", 0)
        svr = col.get("sentinelValueRatio", 0)
        try: nr = float(nr)
        except: nr = 0
        try: esr = float(esr)
        except: esr = 0
        try: svr = float(svr)
        except: svr = 0
        if nr > QUALITY_WARN_THRESHOLDS["nullRatio"]:
            warnings.append(f"NULL率({nr:.1%})偏高")
        if esr > QUALITY_WARN_THRESHOLDS["emptyStringRatio"]:
            warnings.append(f"空字符串率({esr:.1%})过高")
        if svr > QUALITY_WARN_THRESHOLDS["sentinelValueRatio"]:
            warnings.append(f"哨兵值率({svr:.1%})异常")
        qs = 1.0 - min(nr*0.4, 0.4) - min(esr*0.3, 0.3) - min(svr*0.3, 0.3)
        quality_issues.append({
            "table": table_name, "column": col_name,
            "nullRatio": round(nr,4), "emptyStringRatio": round(esr,4),
            "sentinelValueRatio": round(svr,4), "qualityScore": round(max(0,qs),4),
            "warning": "; ".join(warnings) if warnings else "正常"
        })
    audit_data = {
        "sessionId": session_id or f"bridge_{datetime.datetime.now().strftime('%Y%m%d_%H%M%S')}",
        "timestamp": datetime.datetime.now().isoformat(),
        "title": f"数据质量审计 - {table_name}",
        "skills": skills or ["java-mysql-query [已有]", "java-superpowers-contract [已有]"],
        "tools": tools or ["DatabaseQuery", "SkillBridge"],
        "filesRead": [], "filesModified": [], "sqlExecuted": [],
        "dataQualityIssues": quality_issues,
        "summary": {"totalSkills": 2, "totalTools": 2, "totalFilesRead": 0,
                    "totalFilesModified": 0, "totalSqlExecuted": 0,
                    "totalQualityIssues": len([q for q in quality_issues if q["warning"] != "正常"])}
    }
    return audit_data

def main():
    parser = argparse.ArgumentParser(description="Skill Bridge 技能桥接工具")
    parser.add_argument("--analyze-result", help="DatabaseQuery --analyze-table 的JSON结果文件")
    parser.add_argument("--audit-output", default="audit_input.json", help="输出的审计JSON文件")
    parser.add_argument("--audit-format", choices=["json","markdown","html"], default="markdown")
    parser.add_argument("--output", default="combined_report", help="最终报告文件（不含扩展名）")
    parser.add_argument("--db", help="数据库名（直接运行分析用）")
    parser.add_argument("--tables", nargs="+", help="要分析的表名列表")
    parser.add_argument("--host", default="localhost"); parser.add_argument("--port", type=int, default=3306)
    parser.add_argument("--user", default="root")
    parser.add_argument("--password", default=os.environ.get("DB_PASSWORD",""))
    args = parser.parse_args()

    if args.analyze_result:
        audit_data = convert_analyze_to_audit(args.analyze_result)
        with open(args.audit_output, "w") as f:
            json.dump(audit_data, f, ensure_ascii=False, indent=2)
        # 运行时生成报告
        report_path = f"{args.output}.{args.audit_format}"
        subprocess.run(["python3", "scripts/audit_report_generator.py", "--input", args.audit_output,
                       "--format", args.audit_format, "--output", report_path])
        print(json.dumps({"status":"success","audit_data":args.audit_output,"report":report_path}))
    elif args.db and args.tables:
        # 直接分析表并生成审计报告
        all_issues = []
        q = MySQLQuery(args.host, args.port, args.db, args.user, args.password)
        q.connect()
        for table in args.tables:
            data = q.analyze_table(table)
            audit = convert_analyze_to_audit(data, session_id=f"bridge_{table}")
            all_issues.extend(audit["dataQualityIssues"])
        q.close()
        combined_audit = {
            "sessionId": f"bridge_{datetime.datetime.now().strftime('%Y%m%d_%H%M%S')}",
            "timestamp": datetime.datetime.now().isoformat(),
            "title": f"数据质量审计 - {', '.join(args.tables)}",
            "skills": ["java-mysql-query [已有]", "java-superpowers-contract [已有]"],
            "tools": ["DatabaseQuery", "SkillBridge"],
            "filesRead": [], "filesModified": [], "sqlExecuted": [],
            "dataQualityIssues": all_issues,
            "summary": {"totalSkills": 2, "totalTools": 2, "totalFilesRead": 0,
                        "totalFilesModified": 0, "totalSqlExecuted": 0,
                        "totalQualityIssues": len([q for q in all_issues if q["warning"] != "正常"])}
        }
        with open(args.audit_output, "w") as f:
            json.dump(combined_audit, f, ensure_ascii=False, indent=2)
        report_path = f"{args.output}.{args.audit_format}"
        subprocess.run(["python3", "scripts/audit_report_generator.py", "--input", args.audit_output,
                       "--format", args.audit_format, "--output", report_path])
        print(json.dumps({"status":"success","tables":args.tables,"issues":len(all_issues),"report":report_path}))
    else:
        parser.print_help()
if __name__ == "__main__": main()

File: skills/java-superpowers-contract/scripts/skill-bridge.js
#!/usr/bin/env node
/**
 * Skill Bridge (Node.js 版)
 * 连接 java-mysql-query 与 java-superpowers-contract 的桥接工具。
 * 用法: node skill-bridge.js --db mydb --tables user order --output report.md
 */
const fs = require("fs");
const { execSync } = require("child_process");
const { MySQLQuery } = require("./database-query.js");
const path = require("path");

const QUALITY_WARN = { nullRatio: 0.2, emptyStringRatio: 0.3, sentinelValueRatio: 0.1 };

function convertAnalyzeToAudit(analyzeData, sessionId) {
  const analysis = analyzeData.analysis || analyzeData;
  const columns = analysis.columns || [];
  const tableName = analysis.table || "unknown";
  const qualityIssues = [];

  columns.forEach(col => {
    const warnings = [];
    const nr = parseFloat(col.nullRatio) || 0;
    const esr = parseFloat(col.emptyStringRatio) || 0;
    const svr = parseFloat(col.sentinelValueRatio) || 0;
    if (nr > QUALITY_WARN.nullRatio) warnings.push(`NULL率(${(nr*100).toFixed(1)}%)偏高`);
    if (esr > QUALITY_WARN.emptyStringRatio) warnings.push(`空字符串率(${(esr*100).toFixed(1)}%)过高`);
    if (svr > QUALITY_WARN.sentinelValueRatio) warnings.push(`哨兵值率(${(svr*100).toFixed(1)}%)异常`);
    const qs = Math.max(0, 1 - Math.min(nr*0.4,0.4) - Math.min(esr*0.3,0.3) - Math.min(svr*0.3,0.3));
    qualityIssues.push({
      table: tableName, column: col.name || "unknown",
      nullRatio: Math.round(nr*10000)/10000,
      emptyStringRatio: Math.round(esr*10000)/10000,
      sentinelValueRatio: Math.round(svr*10000)/10000,
      qualityScore: Math.round(qs*10000)/10000,
      warning: warnings.length > 0 ? warnings.join("; ") : "正常"
    });
  });

  return {
    sessionId: sessionId || `bridge_${Date.now()}`,
    timestamp: new Date().toISOString(),
    title: `数据质量审计 - ${tableName}`,
    skills: ["java-mysql-query [已有]", "java-superpowers-contract [已有]"],
    tools: ["DatabaseQuery", "SkillBridge"],
    filesRead: [], filesModified: [], sqlExecuted: [],
    dataQualityIssues: qualityIssues,
    summary: {
      totalSkills: 2, totalTools: 2, totalFilesRead: 0,
      totalFilesModified: 0, totalSqlExecuted: 0,
      totalQualityIssues: qualityIssues.filter(q => q.warning !== "正常").length
    }
  };
}

function parseArgs() {
  const args = process.argv.slice(2);
  const opts = { host: "localhost", port: "3306", user: "root", auditFormat: "markdown", output: "combined_report" };
  for (let i = 0; i < args.length; i++) {
    switch (args[i]) {
      case "--analyze-result": opts.analyzeResult = args[++i]; break;
      case "--db": opts.db = args[++i]; break;
      case "--tables": opts.tables = []; while (i+1 < args.length && !args[i+1].startsWith("--")) opts.tables.push(args[++i]); break;
      case "--audit-output": opts.auditOutput = args[++i]; break;
      case "--audit-format": opts.auditFormat = args[++i]; break;
      case "--output": opts.output = args[++i]; break;
      case "--host": opts.host = args[++i]; break;
      case "--port": opts.port = args[++i]; break;
      case "--user": opts.user = args[++i]; break;
      case "--password": opts.password = args[++i]; break;
    }
  }
  return opts;
}

function run(opts) {
  if (opts.analyzeResult) {
    const raw = fs.readFileSync(opts.analyzeResult, "utf-8");
    const analyzeData = JSON.parse(raw);
    const auditData = convertAnalyzeToAudit(analyzeData);
    const auditPath = opts.auditOutput || "audit_input.json";
    fs.writeFileSync(auditPath, JSON.stringify(auditData, null, 2));
    const reportPath = `${opts.output}.${opts.auditFormat}`;
    execSync(`node scripts/audit-report-generator.js --input "${auditPath}" --format ${opts.auditFormat} --output "${reportPath}"`, { shell: true });
    console.log(JSON.stringify({ status: "success", auditData: auditPath, report: reportPath }));
  } else if (opts.db && opts.tables) {
    const allIssues = [];
    opts.tables.forEach(table => {
      const cmd = `python database_query.py --host ${opts.host} --port ${opts.port} --db ${opts.db} --user ${opts.user}${opts.password ? " --password " + opts.password : ""} --analyze-table ${table}`;
      const out = execSync(cmd, { timeout: 120000, shell: true }).toString();
      const data = JSON.parse(out);
      const audit = convertAnalyzeToAudit(data, `bridge_${table}`);
      allIssues.push(...audit.dataQualityIssues);
    });
    const combined = {
      sessionId: `bridge_${Date.now()}`, timestamp: new Date().toISOString(),
      title: `数据质量审计 - ${opts.tables.join(", ")}`,
      skills: ["java-mysql-query [已有]", "java-superpowers-contract [已有]"],
      tools: ["DatabaseQuery", "SkillBridge"], filesRead: [], filesModified: [], sqlExecuted: [],
      dataQualityIssues: allIssues,
      summary: { totalSkills: 2, totalTools: 2, totalFilesRead: 0, totalFilesModified: 0,
                 totalSqlExecuted: 0, totalQualityIssues: allIssues.filter(q => q.warning !== "正常").length }
    };
    const auditPath = opts.auditOutput || "audit_input.json";
    fs.writeFileSync(auditPath, JSON.stringify(combined, null, 2));
    const reportPath = `${opts.output}.${opts.auditFormat}`;
    execSync(`node scripts/audit-report-generator.js --input "${auditPath}" --format ${opts.auditFormat} --output "${reportPath}"`, { shell: true });
    console.log(JSON.stringify({ status: "success", tables: opts.tables, issues: allIssues.length, report: reportPath }));
  } else {
    console.log("用法: --analyze-result <file> | --db <name> --tables <t1> [t2 ...]");
  }
}
run(parseArgs());
